product/util.py

```diff
@@ -171,164 +211,197 @@ def diff_products(left_product, right_product, ignore_component_names=()):
-            if c.name() == changed_component.name():
+            if c.name == changed_component.name:
                 return (changed_component, c)
         return (changed_component, None) # no pair component found
 
-    components_with_changed_versions = FluentIterable(items=components_only_left) \
+    components_with_changed_versions = FluentIterable(items=left_components) \
         .map(functools.partial(find_changed_component, components=right_components)) \
         .filter(lambda cs: cs[1] is not None) \
         .as_list()
-    # pairs of crefs (left-version:right-version)
+    # pairs of components (left:right-version)
 
-    left_names = set(map(lambda c: c.name(), components_only_left))
-    right_names = set(map(lambda c: c.name(), components_only_right))
-    names_version_changed = set(map(lambda cp: cp[0].name(), components_with_changed_versions))
+    left_names = {i.name for i in left_identities}
+    right_names = {i.name for i in right_identities}
+    names_version_changed = {c[0].name for c in components_with_changed_versions}
 
     both_names = left_names & right_names
     left_names -= both_names
     right_names -= both_names
 
     return ComponentDiff(
-        crefs_only_left=components_only_left,
-        crefs_only_right=components_only_right,
-        crefpairs_version_changed=set(components_with_changed_versions),
+        cidentities_only_left=left_only_identities,
+        cidentities_only_right=right_only_identities,
+        cpairs_version_changed=components_with_changed_versions,
         names_only_left=left_names,
         names_only_right=right_names,
         names_version_changed=names_version_changed,
     )
 
 
 @dataclasses.dataclass
-class ImageDiff:
-    left_component: Component
-    right_component: Component
-    irefs_only_left: set = dataclasses.field(default_factory=set)
-    irefs_only_right: set = dataclasses.field(default_factory=set)
-    irefpairs_version_changed: set = dataclasses.field(default_factory=set)
-    names_only_left: set = dataclasses.field(default_factory=set)
-    names_only_right: set = dataclasses.field(default_factory=set)
-    names_version_changed: set = dataclasses.field(default_factory=set)
-
-
-def diff_images(
-    left_component_descriptor,
-    right_component_descriptor,
-    left_component,
-    right_component
-):
-    left_images = set(_effective_images(left_component_descriptor, left_component))
-    right_images = set(_effective_images(right_component_descriptor, right_component))
+class ResourceDiff:
+    left_component: cm.Component
+    right_component: cm.Component
+    resource_refs_only_left: typing.List[cm.Resource] = dataclasses.field(default_factory=list)
+    resource_refs_only_right: typing.List[cm.Resource] = dataclasses.field(default_factory=list)
+    resourcepairs_version_changed: typing.List[typing.Tuple[cm.Resource, cm.Resource]] = dataclasses.field(default_factory=list) # noqa:E501
+
+
+def _add_if_not_duplicate(list, res):
```

Comment:
if only there were some datastructure that would feature only allowing unique elements ^^